### PR TITLE
Validate website field

### LINKED ISSUE #1087: add new checkup: validate that website is not just a link to github repo or to pypi package site
[Sometimes](https://github.com/Qiskit/ecosystem/pull/1059/changes#r3060590054) new members suggest the github README as a website home page for the project. The field `website` is not mandatory and there is no need to point out again to one of the existing links in the submission. It would be great to add a checkup for that.

### PR DESCRIPTION
**Pull request by AdwaithaV**
This PR adds a validation check to prevent redundant website URLs in submissions. The new validation rejects cases where member.website duplicates an existing GitHub repository URL or a PyPI project URL.

@1ucian0 @garrison @blakejohnson I also attempted to run the full local pytest suite, but test collection currently fails due to an unrelated existing syntax error in tests/test_check.py on the main branch. My changes are limited to ecosystem/validation/test_url.py.

Please let me know if this validation approach looks acceptable. If needed, I can also open a separate issue and work on fixing the pytest syntax error in tests/test_check.py as a follow-up contribution.

Fixes #1087

### CODE REVIEW

**Comment by AdwaithaV** [comment]
@1ucian0 @garrison @blakejohnson  I tested the changes locally using pytest, pylint, and black formatting checks. The validation tests passed successfully, pylint reported a 10.00/10 score, and the file was reformatted using black to match repository style requirements.
image
image
image

**Comment by AdwaithaV** [comment]
This is the latest update and final version of the PR. I improved the validation logic by normalizing hostnames before comparison, added handling for repository subpaths so URLs like `/tree/main` and other nested GitHub paths are correctly detected, and added support for additional PyPI hostname variants such as `pypi.python.org`. I also expanded the documentation URL validation checks to cover more GitHub documentation edge cases. The updated file was formatted using `black`, verified with `pylint`, and successfully checked with `python -m py_compile`. The targeted validation tests are passing successfully. Running the complete `pytest` suite is still blocked by an existing upstream syntax error in `tests/test_check.py`, along with unrelated missing optional dependencies in some `documentation/` directories. @garrison @1ucian0 @blakejohnson Please review the changes whenever convenient, and let me know if any further improvements are needed.

**Comment by 1ucian0** [comment]
No need to ping at people, the PR will get reviewed eventually :)

**Comment by AdwaithaV** [comment]
Thank you for letting me know. Sorry I didn't know about this. I am still learning how open source works. I won't ping reviewers in future PRs. Thanks for the guidance..

File: ecosystem/validation/test_url.py [hunk 1]
```diff
@@ -29,33 +31,128 @@ def get_all_urls(cls, member):
...
             for suffix in suffixes:
                 assert not documentation_url.path.endswith(suffix), (
-                    "The field `member.documentation` can be empty. "
-                    "It does not have to be a link to the `README.md`."
+                    "The field `member.documentation` can be "
+                    "empty. It does not have to be a link to "
+                    "the `README.md` or repository root."
+                )
+
+    def test_027(self, member):
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py:81]: you need to extend `checks.toml` with this checkup.

File: ecosystem/validation/test_url.py [hunk 2]
```diff
@@ -29,33 +31,128 @@ def get_all_urls(cls, member):
...
+                "tree/main",
+            ]
+
             for suffix in suffixes:
                 assert not documentation_url.path.endswith(suffix), (
-                    "The field `member.documentation` can be empty. "
-                    "It does not have to be a link to the `README.md`."
+                    "The field `member.documentation` can be "
+                    "empty. It does not have to be a link to "
+                    "the `README.md` or repository root."
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]:
```suggestion
                    "empty. It does not have to be a link to "
                    "the repository's `README.md` or root."
```

File: ecosystem/validation/test_url.py [hunk 3]
```diff
@@ -29,33 +31,128 @@ def get_all_urls(cls, member):
             else:
                 continue
 
+    @staticmethod
+    def normalize_url(url):
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]: Maybe you can bake these normalizatoins inside `get_all_urls`

File: ecosystem/validation/test_url.py [hunk 4]
```diff
@@ -29,33 +31,128 @@ def get_all_urls(cls, member):
...
+        # Reject PyPI URLs as website
+        pypi_packages = getattr(member, "pypi", {})
+
+        for _, package in pypi_packages.items():
+            package_url = getattr(package, "url", None)
+
+            if package_url is None:
+                continue
+
+            hostname = (package_url.hostname or "").lower()
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]: I think URL already lowers the case of hostname

File: ecosystem/validation/test_url.py [hunk 5]
```diff
@@ -29,33 +31,128 @@ def get_all_urls(cls, member):
...
+                repository_url,
+            ), (
+                "The field `member.website` should not "
+                "duplicate the GitHub repository URL."
+            )
+
+        # Reject PyPI URLs as website
+        pypi_packages = getattr(member, "pypi", {})
+
+        for _, package in pypi_packages.items():
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]: no need to fetch the python package. If the website is a pypi link, that's enough.

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
Thanks! few comments here and there.

File: ecosystem/validation/test_url.py (hunk 3 above)
Comment (AdwaithaV) [reply ecosystem/validation/test_url.py]: Thaks for the suggestion .Since `get_all_urls()` is also used by `test_http()`, would you prefer inlining the normalization into` is_same_or_subpath()`, or did you have a different approach in mind for `get_all_urls()`

File: ecosystem/validation/test_url.py [hunk 6]
```diff
@@ -29,33 +31,114 @@ def get_all_urls(cls, member):
...
+
+        if hostname.startswith("www."):
+            hostname = hostname[4:]
+
+        path = parsed.path.rstrip("/")
+
+        return hostname, path
+
+    @staticmethod
+    def is_same_or_subpath(candidate, reference):
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]: Maybe it can be implemented as an assertion:
```suggestion
    def assertSameSubpath(self, candidate, reference):
```

File: resources/checks.toml [hunk 7]
```diff
@@ -154,6 +154,15 @@ checker = "test_url.py::TestURLs::test_026"
 category = "SUBMISSION"
 importance = "RECOMMENDATION"
 
+[027]
+title = "Website should not duplicate GitHub or PyPI URLs"
```
Comment (1ucian0) [inline resources/checks.toml]:
```suggestion
title = "Website should not be GitHub repository or PyPI URL"
```

File: resources/checks.toml [hunk 8]
```diff
@@ -154,6 +154,15 @@ checker = "test_url.py::TestURLs::test_026"
...
+description = "The `member.website` field should point to the project's website instead of duplicating the GitHub repository or a PyPI project page."
```
Comment (1ucian0) [inline resources/checks.toml]:
```suggestion
description = "The `member.website` field should point to the project's website instead of the GitHub repository or a PyPI project page."
```

File: resources/checks.toml [hunk 9]
```diff
@@ -154,6 +154,15 @@ checker = "test_url.py::TestURLs::test_026"
...
+applies_to = "all"
+affects = ["member.website"]
+checker = "test_url.py::TestURLs::test_027"
+category = "SUBMISSION"
+importance = "RECOMMENDATION"
```
Comment (1ucian0) [inline resources/checks.toml]:
```suggestion
importance = "STRONG-RECOMMENDATION"
```

File: ecosystem/validation/test_url.py [hunk 10]
```diff
@@ -29,33 +31,114 @@ def get_all_urls(cls, member):
...
+                website_url,
+                repository_url,
+            ), (
+                "The field `member.website` should not "
+                "duplicate the GitHub repository URL."
+            )
+
+        assert not (
+            hostname.endswith("pypi.org") or hostname.endswith("pypi.python.org")
+        ), ("The field `member.website` should not " "point to a PyPI project URL.")
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]:
```suggestion
        ), ("The field `member.website` should not point to a PyPI project URL.")
```

File: ecosystem/validation/test_url.py [hunk 11]
```diff
@@ -29,33 +31,114 @@ def get_all_urls(cls, member):
...
+
+        if repository_url is not None and repository_url.hostname.lower().endswith(
+            "github.com"
+        ):
+            assert not TestURLs.is_same_or_subpath(
+                website_url,
+                repository_url,
+            ), (
+                "The field `member.website` should not "
+                "duplicate the GitHub repository URL."
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]:
```suggestion
                "The field `member.website` should not "
                "be the GitHub repository."
```

File: ecosystem/validation/test_url.py [hunk 12]
```diff
@@ -29,33 +31,114 @@ def get_all_urls(cls, member):
...
+
+        website_url = getattr(member, "website")
+
+        if website_url is None:
+            pytest.skip("No member.website")
+
+        hostname = website_url.hostname or ""
+
+        # Reject GitHub repository URLs as website
+        repository_url = getattr(member, "url")
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py]: I think there is no need to fetch member.url. A `member.website` that refers to a github repo is reason enough to fail.

File: ecosystem/validation/test_url.py [hunk 13]
```diff
@@ -20,7 +22,7 @@
 class TestURLs:
     @classmethod
     def get_all_urls(cls, member):
-        """recursevly search for URLs in the member data"""
+        """recursively search for URLs in the member data"""
```
Comment (1ucian0) [inline ecosystem/validation/test_url.py:23]: good catch.

File: ecosystem/validation/test_url.py (hunk 3 above)
Comment (1ucian0) [reply ecosystem/validation/test_url.py]: maybe in `is_same_or_subpath`?

**Review by 1ucian0 — COMMENTED** [review]
I just realized that I had some pending comments, not sure if they are sitll current.

**Comment by AdwaithaV** [comment]
I have addressed the remaining review comments. The validation now checks the `member.website` hostname directly for GitHub and PyPI URLs without fetching `member.url`. I also removed the URL normalization and subpath comparison helpers because they are no longer required. The updated file was formatted with Black and passed the syntax check.

**Review by 1ucian0 — COMMENTED** [review]
thanks!

**Review by 1ucian0 — APPROVED** [review]
